src/kernel/bitcoinkernel.h

```diff
@@ -0,0 +1,1142 @@
+
+/**
+ * Opaque data structure for holding a chainstate manager.
+ *
+ * The chainstate manager is the central object for doing validation tasks as
+ * well as retrieving data from the chain. Internally it is a complex data
+ * structure with diverse functionality.
+ *
+ * The chainstate manager is only valid for as long as the context with which it
+ * was created remains in memory.
+ *
+ * Its functionality will be more and more exposed in the future.
+ */
+typedef struct kernel_ChainstateManager kernel_ChainstateManager;
+
+/**
+ * Opaque data structure for holding parameters used for loading the chainstate
+ * of a chainstate manager.
+ *
+ * Is initialized with default parameters.
+ */
+typedef struct kernel_ChainstateLoadOptions kernel_ChainstateLoadOptions;
+
+/**
+ * Opaque data structure for holding a block.
+ */
+typedef struct kernel_Block kernel_Block;
+
+/**
+ * Opaque data structure for holding a non-owned block. This is typically a
+ * block available to the user through one of the validation callbacks.
+ */
+typedef struct kernel_BlockPointer kernel_BlockPointer;
+
+/**
+ * Opaque data structure for holding the state of a block during validation.
+ *
+ * Contains information indicating whether validation was successful, and if not
+ * which step during block validation failed.
+ */
+typedef struct kernel_BlockValidationState kernel_BlockValidationState;
+
+/**
+ * Opaque data structure for holding a validation interface.
+ *
+ * The validation interface can be registered with the internal task runner of an
+ * existing context. It holds callbacks that will be triggered by certain
+ * validation events. The callbacks registered with it will block further
+ * validation progress when triggered.
+ */
+typedef struct kernel_ValidationInterface kernel_ValidationInterface;
+
+/**
+ * Opaque data structure for holding a block undo struct.
+ *
+ * It holds all the previous outputs consumed by all transactions in a specific
+ * block. Internally it holds a nested vector. The top level vector has an entry
+ * for each transaction in a block (in order of the actual transactions of the
+ * block and minus the coinbase transaction). Each entry is in turn a vector of
+ * all the previous outputs of a transaction (in order of their corresponding
+ * inputs).
+ */
+typedef struct kernel_BlockUndo kernel_BlockUndo;
+
+/** Current sync state passed to tip changed callbacks. */
+typedef enum {
+    kernel_INIT_REINDEX,
+    kernel_INIT_DOWNLOAD,
+    kernel_POST_INIT
+} kernel_SynchronizationState;
+
+/** Possible warning types issued by validation. */
+typedef enum {
+    kernel_UNKNOWN_NEW_RULES_ACTIVATED,
+    kernel_LARGE_WORK_INVALID_CHAIN
+} kernel_Warning;
+
+/** Callback function types */
+
+/**
+ * Function signature for the global logging callback. All bitcoin kernel
+ * internal logs will pass through this callback.
+ */
+typedef void (*kernel_LogCallback)(void* user_data, const char* message);
```

Comment:
A general comment on the API: i'd prefer to pass (and receive) explicit lengths for strings instead of bare `char*`s.

My experience with wrapping C APIs in rust is that it's important to have a defined memory-range for strings and arrays. Relying on NUL-termination means that the memory size is effectively unrestricted, making it impossible to do some checks safely. This is (with lesser urgency) also true for other languages like Python that represent strings as pointer+length.

Reply:
Mmh, thanks for this. It should be easy to add a length parameter here.